service/qos/service_level_controller.cc

```diff
@@ -410,4 +410,16 @@ future<> service_level_controller::do_remove_service_level(sstring name, bool re
     return make_ready_future();
 }
 
+void service_level_controller::on_join_cluster(const gms::inet_address& endpoint, bool this_node) { }
+
+void service_level_controller::on_leave_cluster(const gms::inet_address& endpoint, bool this_node) {
+    if (this_node && this_shard_id() == global_controller) {
```

Comment:
So, request_abort is called either when a node is decommissioned or the node stops, right? Why not just stop it like we stop other things (e.g., ss.stop_transport()) in storage_service::decommission.

Reply:
I think that this coupling is unhealthy. It means that storage_service needs to know about every module that is dependent upon endpoints (or at least the local endpoint) life-cycle.
I don't have anything about doing it this way, I simply think it will strongly couple the `service_level_controller` and `storage_service` together. But if you think this is the best approach, I don't mind.

Reply:
> So, request_abort is called either when a node is decommissioned or the node stops, right? Why not just stop it like we stop other things (e.g., ss.stop_transport()) in storage_service::decommission.

Here I took your advice, I will need to stop this process for other cases of leaving the cluster too :slightly_smiling_face:

Reply:
Hi @eliransin, we can just pass the ip address to service_level_controller when we construct service_level_controller to avoid coupling with stotrage_service. I do not like coupling them together neither ;)

Reply:
> Hi @eliransin, we can just pass the ip address to service_level_controller when we construct service_level_controller to avoid coupling with stotrage_service. I do not like coupling them together neither ;)

I have actually noticed that there is a utility function (which storage_service uses) that gives you the broadcast address.
Another version will come soon (I am testing it).